cpp/src/fil/treelite_import.cu

```diff
@@ -122,35 +151,27 @@ template <typename T, typename L>
 inline std::vector<cat_feature_counters> cat_counter_vec(const tl::Tree<T, L>& tree, int n_cols)
 {
   std::vector<cat_feature_counters> res(n_cols);
-  std::stack<int> stack;
-  stack.push(tree_root(tree));
-  while (!stack.empty()) {
-    int node_id = stack.top();
-    stack.pop();
-    while (!tree.IsLeaf(node_id)) {
-      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical) {
-        std::vector<std::uint32_t> mmv = tree.MatchingCategories(node_id);
-        int max_matching_cat;
-        if (mmv.size() > 0) {
-          // in `struct cat_feature_counters` and GPU structures, max matching category is an int
-          // cast is safe because all precise int floats fit into ints, which are asserted to be 32
-          // bits
-          max_matching_cat = mmv.back();
-          ASSERT(max_matching_cat <= MAX_FIL_INT_FLOAT,
-                 "FIL cannot infer on "
-                 "more than %d matching categories",
-                 MAX_FIL_INT_FLOAT);
-        } else {
-          max_matching_cat = -1;
-        }
-        cat_feature_counters& counters = res[tree.SplitIndex(node_id)];
-        counters =
-          cat_feature_counters::combine(counters, cat_feature_counters{max_matching_cat, 1});
+  walk_tree<empty>(tree, [&](int node_id, empty val) {
```

Comment:
`empty val` clearly looks redundant (though it's necessary for the current solution).

Consider providing a wrapper version of `walk_tree` which accepts functions without additional dummy arguments.

Alternatively, you could use the previous version. It's definitely better for the function to be more readable at the call site, the definition is going to be heavily templatized anyway.